src/index.ts

```diff
@@ -0,0 +1,223 @@
+import dotenv from "dotenv";
+import express, { Express, Request, Response } from "express";
+import { getIronSession, IronSession } from 'iron-session';
+import { BaseClient, Issuer, generators } from 'openid-client';
+import { getClientIp } from 'request-ip';
+import crypto from 'crypto';
+
+dotenv.config();
+
+const app: Express = express();
+const port = process.env.PORT || 3000;
+const authorizedUsers = (process.env.AUTHORIZED_USER_EMAILS || '').split(',');
+
+// SESSION
+const IRON_SESSION_PWD = (process.env.IRON_SESSION_PWD || '')
+const COOKIE_NAME = (process.env.AUTH_COOKIE_NAME || 'annuaire-entreprises-admin-auth-session')
+const COOKIE_DOMAIN = process.env.AUTH_COOKIE_DOMAIN || '127.0.0.1'
+const SESSION_TTL = Number.parseInt(process.env.AUTH_SESSION_TTL || '3600')
+
+// OIDC
+const CLIENT_ID = process.env.OPENID_CLIENT_ID || '';
+const CLIENT_SECRET = process.env.OPENID_CLIENT_SECRET || '';
+const ISSUER_URL = process.env.OPENID_URL_DISCOVER || '';
+const REDIRECT_URI = process.env.OPENID_REDIRECT_URI || '';
+const POST_LOGOUT_REDIRECT_URI = process.env.OPENID_POST_LOGOUT_REDIRECT_URI || '';
+
+// constants to check if the session has been reused
+const VERIFY_BROWSER_SIGNATURE = process.env.VERIFY_BROWSER_SIGNATURE == '1'
+const VERIFY_IP_ADDRESS = process.env.VERIFY_IP_ADDRESS == '1'
+
+const SCOPES = 'openid email';
+
+let openidClient = undefined as BaseClient | undefined;
+
+type Session = {
+  email?: string;
+
+  ip?: string | null;
+  signature?: string;
+
+  // OIDC Issuer
+  state?: string;
+  nonce?: string;
+  idToken?: string;
+
+  // connexion
+  pathFrom?: string;
+};
+
+type OpenConnectUserInfo = {
+  email: string;
+  email_verified: boolean;
+};
+
+const getSession = async (req: Request, res: Response) => {
+  return await getIronSession<Session>(req, res, {
+    password: IRON_SESSION_PWD,
+    cookieName: COOKIE_NAME,
+    ttl: SESSION_TTL,
+    cookieOptions: {
+      httpOnly: true,
+      secure: true,
+      sameSite: "lax",
+      maxAge: SESSION_TTL - 60,
+      domain: COOKIE_DOMAIN,
+    }
+  });
+}
+
+const getOpenIDClient = async () => {
+  if (openidClient) {
+    return openidClient;
+  } else {
+    const issuer = await Issuer.discover(ISSUER_URL);
+
+    openidClient = new issuer.Client({
+      client_id: CLIENT_ID,
+      client_secret: CLIENT_SECRET,
+      redirect_uris: [REDIRECT_URI],
+      post_logout_redirect_uris: [POST_LOGOUT_REDIRECT_URI],
+      userinfo_signed_response_alg: 'RS256',
+    });
+
+    return openidClient;
+  }
+};
+
+const getBrowserSignature = (req: Request) => {
+  const headers = [
+    'user-agent',
+    'accept',
+    'accept-language',
+    'accept-encoding',
+  ]
+
+  const raw = headers.map(header => req.headers[header] || '').join('\n')
+  const signature = crypto.createHash('md5').update(raw).digest('hex')
+
+  return signature
+}
```

Comment:
@johangirod this is very interesting. We should add this to our antibot logic on website